### PR TITLE
[Model] Support DP for ViT on Kimi-VL-A3B-Thinking-2506

### PR DESCRIPTION
**Pull request by david6666666**
## Purpose
Add option to run MiniCPM-V-4 vision encoder in data parallel manner while the main model is in TP. Can be enabled by flag: mm_encoder_tp_mode="data"
## Test Plan
```
export CUDA_VISIBLE_DEVICES=4,5,6,7
vllm serve \
    /workspace/models/Kimi-VL-A3B-Thinking-2506 \
    --port 5580 --host 0.0.0.0 \
    --max-num-seqs 128 --dtype bfloat16 --max-model-len=8192 \
    --no-enable-prefix-caching --trust-remote-code -tp 4 \
    --allowed-local-media-path /workspace/l00807937/ \
    --gpu-memory-utilization=0.93 \
    --mm-encoder-tp-mode data
```

```
curl -X POST  http://127.0.0.1:5580/v1/chat/completions \
    -H "Content-Type: application/json" \
    -d '{
    "model": "/workspace/models/Kimi-VL-A3B-Thinking-2506",
    "messages": [
    {"role": "system", "content": "You are a helpful assistant."},
    {"role": "user", "content": [
        {"type": "image_url", "image_url": {"url": "file:///workspace/l00807937/eagle_mm/image/GGBond.jpg"}},
        {"type": "text", "text": "What is the text in the illustrate?"}
    ]}
    ]
    }'
```
## Test Result
banchmark
```
baseline
============ Serving Benchmark Result ============
Successful requests:                     99        
Benchmark duration (s):                  10.73     
Total input tokens:                      11015     
Total generated tokens:                  12646     
Request throughput (req/s):              9.22      
Output token throughput (tok/s):         1178.11   
Total Token throughput (tok/s):          2204.27   
---------------Time to First Token----------------
Mean TTFT (ms):                          3795.76   
Median TTFT (ms):                        3401.48   
P99 TTFT (ms):                           8197.40   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          51.70     
Median TPOT (ms):                        54.61     
P99 TPOT (ms):                           74.09     
---------------Inter-token Latency----------------
Mean ITL (ms):                           51.58     
Median ITL (ms):                         21.05     
P99 ITL (ms):                            350.18    
==================================================

DP4
============ Serving Benchmark Result ============
Successful requests:                     99        
Benchmark duration (s):                  8.45      
Total input tokens:                      11015     
Total generated tokens:                  12636     
Request throughput (req/s):              11.72     
Output token throughput (tok/s):         1495.34   
Total Token throughput (tok/s):          2798.85   
---------------Time to First Token----------------
Mean TTFT (ms):                          2753.60   
Median TTFT (ms):                        2368.16   
P99 TTFT (ms):                           6068.15   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          42.08     
Median TPOT (ms):                        44.88     
P99 TPOT (ms):                           57.80     
---------------Inter-token Latency----------------
Mean ITL (ms):                           41.74     
Median ITL (ms):                         20.42     
P99 ITL (ms):                            196.15    
==================================================
```
---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: vllm/model_executor/models/kimi_vl.py [hunk 1]
```diff
@@ -292,10 +312,16 @@ def __init__(
         quant_config = vllm_config.quant_config
 
         assert isinstance(config.vision_config, MoonViTConfig)
+        self.use_data_parallel = model_config.multimodal_config.mm_encoder_tp_mode == "data"
```
Comment (DarkLight1337) [inline vllm/model_executor/models/kimi_vl.py:305]: Please apply #23733

File: vllm/model_executor/models/kimi_vl.py [hunk 2]
```diff
@@ -93,29 +95,47 @@ class MaxImageTokenMeta:
...
+                                    prefix=maybe_prefix(prefix, "linear_1"))
+            self.linear_2 = ReplicatedLinear(self.hidden_size,
+                                  config.text_config.hidden_size,
+                                  bias=True,
+                                  prefix=maybe_prefix(prefix, "linear_2"))
+        else:
+            self.linear_1 = nn.Linear(self.hidden_size,
+                                    self.hidden_size,
+                                    bias=True)
+            self.linear_2 = nn.Linear(self.hidden_size,
```
Comment (Isotr0py) [inline vllm/model_executor/models/kimi_vl.py]:
```suggestion
        self.linear_1 = ReplicatedLinear(self.hidden_size,
                                self.hidden_size,
                                bias=True,
                                return_bias=False,
                                prefix=maybe_prefix(prefix, "linear_1"))
        self.linear_2 = ReplicatedLinear(self.hidden_size,
                              config.text_config.hidden_size,
                              bias=True,
                              return_bias=False,
                              prefix=maybe_prefix(prefix, "linear_2"))
```
I think we can directly use `ReplicatedLinear` here, because it's equivalent to `nn.Linear` .

File: vllm/model_executor/models/moonvit.py [hunk 3]
```diff
@@ -383,21 +385,43 @@ class MLP2(nn.Module):
...
+                                bias=bias,
+                                prefix=maybe_prefix(prefix, "linear_1"))
+            self.fc1 = ReplicatedLinear(
+                                dims[1],
+                                dims[2],
+                                bias=bias,
+                                prefix=maybe_prefix(prefix, "linear_2"))
+        else:
+            self.fc0 = nn.Linear(dims[0], dims[1], bias=bias)
+            self.fc1 = nn.Linear(dims[1], dims[2], bias=bias)
```
Comment (Isotr0py) [inline vllm/model_executor/models/moonvit.py]: ditto.

File: vllm/model_executor/models/moonvit.py [hunk 4]
```diff
@@ -423,9 +449,23 @@ def __init__(
...
+                                 hidden_dim * 3,
+                                 bias=attn_bias,
+                                 prefix=f"{prefix}.wqkv")
+            self.wo = ReplicatedLinear(hidden_dim,
+                                hidden_dim,
+                                bias=attn_bias,
+                                prefix=f"{prefix}.wo")
+        else:
+            self.wqkv = nn.Linear(hidden_dim, hidden_dim * 3, bias=attn_bias)
+            self.wo = nn.Linear(hidden_dim, hidden_dim, bias=attn_bias)
```
Comment (Isotr0py) [inline vllm/model_executor/models/moonvit.py]: ditto.

File: vllm/model_executor/models/kimi_vl.py (hunk 1 above)
Comment (LJH-LBJ) [reply vllm/model_executor/models/kimi_vl.py:305]: Done

File: vllm/model_executor/models/kimi_vl.py (hunk 2 above)
Comment (LJH-LBJ) [reply vllm/model_executor/models/kimi_vl.py]: Done

File: vllm/model_executor/models/moonvit.py (hunk 4 above)
Comment (LJH-LBJ) [reply vllm/model_executor/models/moonvit.py]: Done

File: vllm/model_executor/models/moonvit.py (hunk 3 above)
Comment (LJH-LBJ) [reply vllm/model_executor/models/moonvit.py]: Done

**Comment by LJH-LBJ** [comment]
@Isotr0py @DarkLight1337 please check again

File: vllm/multimodal/utils.py [hunk 5]
```diff
@@ -444,7 +444,7 @@ def run_dp_sharded_vision_model(image_input: torch.Tensor,
     Args:
         image_input (torch.Tensor): Image input tensor.
         vision_model (torch.nn.Module): Vision model.
-
+
```
Comment (DarkLight1337) [inline vllm/multimodal/utils.py]: Unnecessary whitespace

File: vllm/multimodal/utils.py [hunk 6]
```diff
@@ -469,6 +469,174 @@ def run_dp_sharded_vision_model(image_input: torch.Tensor,
     return vision_embeddings
 
 
+def run_dp_sharded_mrope_vision_model_tensor_Kimi(
```
Comment (DarkLight1337) [inline vllm/multimodal/utils.py]: Can you explain the difference between the regular mrope method and this one? It seems to me that the main difference is just in the attribute names of `vision_model`

File: vllm/multimodal/utils.py (hunk 5 above)
Comment (LJH-LBJ) [reply vllm/multimodal/utils.py]: Done

File: vllm/multimodal/utils.py (hunk 6 above)
Comment (LJH-LBJ) [reply vllm/multimodal/utils.py]: Besides the attribute names of vision_model, the type of grid_thw and out_embeddings is different

Comment (DarkLight1337) [reply vllm/multimodal/utils.py]: In that case, can you refactor the original mrope method to allow specifying the attribute names? This could avoid a lot of duplicate code

Comment (LJH-LBJ) [reply vllm/multimodal/utils.py]: I will try

File: docs/configuration/optimization.md [hunk 7]
```diff
@@ -178,6 +178,7 @@ Known supported models:
 - MiniCPM-V-4 (<gh-pr:23327>)
 - Qwen2.5-VL (<gh-pr:22742>)
 - Step3 (<gh-pr:22697>)
+- Kimi (<gh-pr:23817>)
```
Comment (DarkLight1337) [inline docs/configuration/optimization.md]: Can you keep this in alphabetical order?

Comment (LJH-LBJ) [reply docs/configuration/optimization.md]: Sure

File: vllm/multimodal/utils.py (hunk 6 above)
Comment (LJH-LBJ) [reply vllm/multimodal/utils.py]: Done, please check again

File: docs/configuration/optimization.md (hunk 7 above)
Comment (LJH-LBJ) [reply docs/configuration/optimization.md]: Done, please check again

**Comment by LJH-LBJ** [comment]
@Isotr0py please check again

File: vllm/multimodal/utils.py [hunk 8]
```diff
@@ -617,23 +621,42 @@ def run_dp_sharded_mrope_vision_model(
     local_grid_thw_list = [grid_thw_list[i] for i in image_idxs_local]
 
     # Run the vision model on the local pixel_values_local
-    if pixel_values_local.shape[0] > 0:
-        image_embeds_local = vision_model(pixel_values_local,
-                                          local_grid_thw_list)
+    if vision_model_name == "MoonVitPretrainedModel":
```
Comment (DarkLight1337) [inline vllm/multimodal/utils.py]: Better not hardcode this inside the method. Perhaps we should add additional arguments to this function to define how to get `image_embeds_local`, `padding` and `out_embeddings`.

Comment (LJH-LBJ) [reply vllm/multimodal/utils.py]: I  think it could be a litte bit hard, because kimi needs different processing with inputs and outputs in this function from other models. If you do not want to see the func name, maybe I can add an attribute in MoonVitPretrainedModel. And then use the attribute distinguish Kimi from other models.

Comment (DarkLight1337) [reply vllm/multimodal/utils.py]: > maybe I can add an attribute in MoonVitPretrainedModel. And then use the attribute distinguish Kimi from other models.

That could work as well

Comment (LJH-LBJ) [reply vllm/multimodal/utils.py]: Done

File: vllm/multimodal/utils.py [hunk 9]
```diff
@@ -605,8 +604,12 @@ def run_dp_sharded_mrope_vision_model(
                                          device=pixel_values.device,
                                          dtype=pixel_values.dtype)
     # embed_dim_reduction_factor = 2 * 2
-    embed_dim_reduction_factor = (vision_model.spatial_merge_size *
-                                  vision_model.spatial_merge_size)
+    if getattr(vision_model, "vit_processing_type", "default") == "rope_2d":
```
Comment (DarkLight1337) [inline vllm/multimodal/utils.py]: Can you add this to the docstring of the function to explain its use?

File: vllm/multimodal/utils.py [hunk 10]
```diff
@@ -674,12 +697,19 @@ def run_dp_sharded_mrope_vision_model(
...
-    assert len(out_embeddings) == len(
-        original_order_embeddings), "Found unassigned embeddings"
-    return out_embeddings
+    if getattr(vision_model, "vit_processing_type", "default") == "rope_2d":
+        out_embeddings_list = [
+            embed for embed in original_order_embeddings if embed is not None
+        ]
+        assert len(out_embeddings_list) == len(
+            original_order_embeddings), "Found unassigned embeddings"
+        return out_embeddings_list
```
Comment (Isotr0py) [inline vllm/multimodal/utils.py]: Why can't we return tuple of embeddings here?

File: vllm/multimodal/utils.py (hunk 9 above)
Comment (LJH-LBJ) [reply vllm/multimodal/utils.py]: Sure， done

File: vllm/multimodal/utils.py [hunk 11]
```diff
@@ -605,8 +604,17 @@ def run_dp_sharded_mrope_vision_model(
                                          device=pixel_values.device,
                                          dtype=pixel_values.dtype)
     # embed_dim_reduction_factor = 2 * 2
-    embed_dim_reduction_factor = (vision_model.spatial_merge_size *
-                                  vision_model.spatial_merge_size)
+    # Special handling for Kimi and models that use 2D rotary position
+    # embedding (rope_2d) for vision processing. By checking
+    # `vit_processing_type`, this logic is extensible: other models that
+    # require rope_2d processing can also be supported in the future by
+    # setting this attribute.
```
Comment (Isotr0py) [inline vllm/multimodal/utils.py]: I think it would be better to move `vit_processing_type` as an argument of this function, because it's not a field in any model's `config.json`:
```python3
def run_dp_sharded_mrope_vision_model(
    vision_model: torch.nn.Module,
    pixel_values: torch.Tensor,
    grid_thw_list: list[list[int]],
    *,
    rope_type: Literal["rope_3d", "rope_2d"],
) -> Union[tuple[torch.Tensor, ...], list[torch.Tensor]]:
```

File: vllm/multimodal/utils.py (hunk 10 above)
Comment (LJH-LBJ) [reply vllm/multimodal/utils.py]: Because in vllm\model_executor\models\kimi_vl.py, it needs the return of  _process_image_pixels, which is the same as the return of run_dp_sharded_mrope_vision_model, to be a list.
``` 
def _process_image_input(self,
                             image_input: KimiVLImageInputs) -> torch.Tensor:
        assert image_input["type"] == "pixel_values"
        image_features = self._process_image_pixels(image_input)
        assert isinstance(image_features, list)
        lengths = [x.shape[0] for x in image_features]
        return self.multi_modal_projector(
            torch.cat(image_features)).split(lengths)
```

Comment (Isotr0py) [reply vllm/multimodal/utils.py]: I see, I think `assert isinstance(image_features, (list, tuple))` should also work.

Comment (LJH-LBJ) [reply vllm/multimodal/utils.py]: Yes, you are right.

File: vllm/multimodal/utils.py (hunk 11 above)
Comment (LJH-LBJ) [reply vllm/multimodal/utils.py]: I let the  output to be type of tuple and add rope_type as an argument of this function. Please check again.

File: vllm/model_executor/models/kimi_vl.py [hunk 12]
```diff
@@ -466,6 +489,30 @@ def compute_logits(self, hidden_states: torch.Tensor,
...
+                if new_name not in qkv_weights:
+                    qkv_weights[new_name] = [None] * 3
+                qkv_weights[new_name][idx] = loaded_weight
+                break
+            else:
+                yield name, loaded_weight
+        for key, weight in qkv_weights.items():
+            qkv_weight = torch.cat(weight, dim=0)
+            yield key, qkv_weight
+
```
Comment (Isotr0py) [inline vllm/model_executor/models/kimi_vl.py]:
```suggestion
```
Seems unused?

Comment (LJH-LBJ) [reply vllm/model_executor/models/kimi_vl.py]: Yes, already deleted

**Comment by LJH-LBJ** [comment]
@Isotr0py Is it good to merge?

File: vllm/model_executor/models/moonvit.py [hunk 13]
```diff
@@ -383,21 +385,35 @@ class MLP2(nn.Module):
...
-        self.fc1 = nn.Linear(dims[1], dims[2], bias=bias)
+        self.use_data_parallel = use_data_parallel
+        self.fc0 = ReplicatedLinear(dims[0],
+                                    dims[1],
+                                    bias=bias,
+                                    prefix=maybe_prefix(prefix, "linear_1"))
+        self.fc1 = ReplicatedLinear(dims[1],
+                                    dims[2],
+                                    bias=bias,
+                                    prefix=maybe_prefix(prefix, "linear_2"))
```
Comment (Isotr0py) [inline vllm/model_executor/models/moonvit.py]:
```suggestion
        self.fc0 = ReplicatedLinear(dims[0],
                                    dims[1],
                                    bias=bias,
                                    prefix=maybe_prefix(prefix, "fc0"))
        self.fc1 = ReplicatedLinear(dims[1],
                                    dims[2],
                                    bias=bias,
                                    prefix=maybe_prefix(prefix, "fc1"))
```

File: vllm/model_executor/models/moonvit.py [hunk 14]
```diff
@@ -383,21 +385,35 @@ class MLP2(nn.Module):
...
         self.activation = activation
         for m in [self.fc0, self.fc1]:
-            nn.init.trunc_normal_(m.weight, std=math.sqrt(2 / m.in_features))
+            nn.init.trunc_normal_(m.weight,
+                                  std=math.sqrt(2 / m.weight.shape[1]))
             if m.bias is not None:
                 nn.init.zeros_(m.bias)
```
Comment (Isotr0py) [inline vllm/model_executor/models/moonvit.py]: Let's remove this `for` loop since it's redundant for inference.

**Review by Isotr0py — APPROVED** [review]
LGTM, just leave some final nits.

File: vllm/model_executor/models/moonvit.py (hunk 14 above)
Comment (LJH-LBJ) [reply vllm/model_executor/models/moonvit.py]: Done

**Comment by LJH-LBJ** [comment]
@DarkLight1337 Please check again

File: docs/configuration/optimization.md [hunk 15]
```diff
@@ -174,6 +174,7 @@ Regardless, you need to set `mm_encoder_tp_mode="data"` in engine arguments to u
 
 Known supported models:
 
+- Kimi (<gh-pr:23817>)
```
Comment (DarkLight1337) [inline docs/configuration/optimization.md]:
```suggestion
- Kimi-VL (<gh-pr:23817>)
```

To be more precise

Comment (LJH-LBJ) [reply docs/configuration/optimization.md]: Sure, Done

**Review by DarkLight1337 — APPROVED** [review]
